Source/WebCore/dom/ElementInternals.cpp

```diff
@@ -50,6 +52,82 @@ ShadowRoot* ElementInternals::shadowRoot() const
     return shadowRoot;
 }
 
+ExceptionOr<RefPtr<HTMLFormElement>> ElementInternals::form() const
+{
+    if (RefPtr element = elementAsFormAssociatedCustom())
+        return element->form();
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<void> ElementInternals::setFormValue(std::optional<CustomElementFormValue>&& value, std::optional<CustomElementFormValue>&& state)
+{
+    if (RefPtr element = elementAsFormAssociatedCustom()) {
+        element->setFormValue(WTFMove(value), WTFMove(state));
+        return { };
+    }
+
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<void> ElementInternals::setValidity(ValidityStateFlags validityStateFlags, String&& message, HTMLElement* anchor)
+{
+    if (RefPtr element = elementAsFormAssociatedCustom())
+        return element->setValidity(validityStateFlags, WTFMove(message), anchor);
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<bool> ElementInternals::willValidate() const
+{
+    if (RefPtr element = elementAsFormAssociatedCustom())
+        return element->willValidate();
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<RefPtr<ValidityState>> ElementInternals::validity()
+{
+    if (RefPtr element = elementAsFormAssociatedCustom())
+        return element->validity();
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<String> ElementInternals::validationMessage() const
+{
+    if (RefPtr element = elementAsFormAssociatedCustom())
+        return element->validationMessage();
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<bool> ElementInternals::reportValidity()
+{
+    if (RefPtr element = elementAsFormAssociatedCustom())
+        return element->reportValidity();
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<bool> ElementInternals::checkValidity()
+{
+    if (RefPtr element = elementAsFormAssociatedCustom())
+        return element->checkValidity();
+    return Exception { NotSupportedError };
+}
+
+ExceptionOr<RefPtr<NodeList>> ElementInternals::labels()
+{
+    if (RefPtr element = dynamicDowncast<HTMLMaybeFormAssociatedCustomElement>(m_element.get())) {
```

Comment:
Why can't we use just use elementAsFormAssociatedCustom here?

Reply:
I see. Either approach is okay to me.

Reply:
Actually there was a bug: `elementAsFormAssociatedCustom` is a bit different from `isFormAssociatedCustomElement` as it allows `ElementInternals` to be used in constructor.

Changed to `elementAsFormAssociatedCustom` and added a test.